tensorstore/kvstore/s3/s3_key_value_store_test.cc

```diff
@@ -0,0 +1,265 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stddef.h>
+
+#include <algorithm>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include "absl/flags/flag.h"
+#include "absl/log/absl_log.h"
+#include "absl/status/status.h"
+#include "absl/strings/match.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
+#include <nlohmann/json.hpp>
+#include "tensorstore/context.h"
+#include "tensorstore/internal/env.h"
+#include "tensorstore/internal/http/curl_handle.h"
+#include "tensorstore/internal/http/curl_transport.h"
+#include "tensorstore/internal/http/http_request.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/json_gtest.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/schedule_at.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/generation.h"
+#include "tensorstore/kvstore/kvstore.h"
+#include "tensorstore/kvstore/operations.h"
+#include "tensorstore/internal/subprocess.h"
+#include "tensorstore/kvstore/test_util.h"
+#include "tensorstore/util/execution/execution.h"
+#include "tensorstore/util/execution/sender_testutil.h"
+#include "tensorstore/util/executor.h"
+#include "tensorstore/util/future.h"
+#include "tensorstore/util/status_testutil.h"
+#include "tensorstore/util/str_cat.h"
+
+namespace kvstore = ::tensorstore::kvstore;
+using ::tensorstore::CompletionNotifyingReceiver;
+using ::tensorstore::Context;
+using ::tensorstore::Future;
+using ::tensorstore::KeyRange;
+using ::tensorstore::MatchesJson;
+using ::tensorstore::MatchesStatus;
+using ::tensorstore::StorageGeneration;
+using ::tensorstore::internal::SpawnSubprocess;
+using ::tensorstore::internal::Subprocess;
+using ::tensorstore::internal::SubprocessOptions;
+using ::tensorstore::TimestampedStorageGeneration;
+using ::tensorstore::StrCat;
+using ::tensorstore::internal::ScheduleAt;
+using ::tensorstore::internal_http::GetDefaultHttpTransport;
+using ::tensorstore::internal_http::HttpRequest;
+using ::tensorstore::internal_http::HttpResponse;
+using ::tensorstore::internal_http::HttpTransport;
+using ::tensorstore::internal_http::HttpRequestBuilder;
+using ::tensorstore::internal_http::SetDefaultHttpTransport;
+using ::tensorstore::internal_kvstore_s3::S3Credentials;
+using ::tensorstore::internal_kvstore_s3::S3RequestBuilder;
+using ::tensorstore::internal::GetEnv;
+using ::tensorstore::internal::SetEnv;
+using ::tensorstore::internal::UnsetEnv;
+
+
+ABSL_FLAG(std::string, localstack_binary, "",
+          "Path to the localstack");
```

Comment:
@laramiel I adapted this from the gcs test bench, but didn't have enough time to understand and investigate how bazel linked (and installed?) the test bench utility. Might this be possible with localstack?

Reply:
Yes, see here:
https://github.com/google/tensorstore/tree/master/third_party/pypa

Add to test_requirements.txt then run generate_workspace.py

Reply:
I'd probably name this s3_localstack_test, then add to the BUILD file `tags = ["manual"]` so it's not triggered on every bazel test ... invocation.

Reply:
... also, split the tests which don't require a running localstack into a separate test.